### PR TITLE
Return unoptimized quote in SwapQuote

### PR DESCRIPTION
**Pull request by xianny**
## Description

Supports new feature: calculating user savings.
Blocks: https://github.com/0xProject/0x-api/pull/435

This PR adds new fields `unoptimizedQuoteInfo` and `unoptimizedOrders` to the SwapQuote. 

These new fields represent the best rate available from a single source, which would be the best alternative to our multi-source optimized path. 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/quote_consumers/utils.ts [hunk 1]
```diff
@@ -31,3 +31,34 @@ export function getSwapMinBuyAmount(quote: SwapQuote): BigNumber {
     }
     return quote.bestCaseQuoteInfo.makerAssetAmount.times(slipRatio).integerValue(BigNumber.ROUND_DOWN);
 }
+
+/**
+ * Same as `getSwapMinBuyAmount` but operates
+ * on a single quote info instead of using best and worst case
+ * Orders must be derived from the same path as the quote info
+ */
+export function getQuoteInfoMinBuyAmount(
```
Comment (xianny) [inline packages/asset-swapper/src/quote_consumers/utils.ts:40]: I use this in 0x-api to calculate affiliate fees if any

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 2]
```diff
@@ -298,28 +290,7 @@ function createTwoHopSwapQuote(
...
-        };
-    } else {
-        return {
-            ...quoteBase,
-            type: MarketOperation.Sell,
-            takerAssetFillAmount: assetFillAmount,
-            makerTokenDecimals,
-            takerTokenDecimals,
-        };
-    }
```
Comment (xianny) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts:322]: Did a little refactoring but functionality should be the same

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 3]
```diff
@@ -148,119 +140,122 @@ export class SwapQuoteCalculator {
...
+    // Calculate the unoptimised alternative
+    const unoptimizedFillResult = simulateBestCaseFill({
         gasPrice,
-        orders: optimizedOrders,
+        orders: unoptimizedPath.orders,
         side: operation,
         fillAmount: assetFillAmount,
         opts: { gasSchedule },
     });
+    const unoptimizedQuoteInfo = fillResultsToQuoteInfo(unoptimizedFillResult);
```
Comment (xianny) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts:189]: This is new, the rest is refactoring

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 4]
```diff
@@ -148,119 +140,122 @@ export class SwapQuoteCalculator {
...
+        takerAssetData,
         gasPrice,
-        bestCaseQuoteInfo: fillResultsToQuoteInfo(bestCaseFillResult),
-        worstCaseQuoteInfo: fillResultsToQuoteInfo(worstCaseFillResult),
-        sourceBreakdown: getSwapQuoteOrdersBreakdown(bestCaseFillResult.fillAmountBySource),
         orders: optimizedOrders,
+        bestCaseQuoteInfo,
+        worstCaseQuoteInfo,
+        unoptimizedQuoteInfo,
+        unoptimizedOrders: unoptimizedPath.orders,
```
Comment (xianny) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts:201]: This line is new too, the rest is refactoring

**Review by dorothy-zbornak — APPROVED** [review]
Yeah, I think this makes sense. Have you hooked it up to API to see if it looks right?

**Comment by xianny** [comment]
Ya, it looks like this in API: https://github.com/0xProject/0x-api/pull/435#issuecomment-736176984

Pretty unambiguous!

**Pull request merged**